### PR TITLE
[mlir][vector] Add support for linearizing Extract, ExtractStridedSlice, Shuffle VectorOps in VectorLinearize

### PR DESCRIPTION
**Pull request by charithaintc**
This PR adds support for converting `vector.extract_strided_slice` and `vector.extract` operations to equivalent `vector.shuffle` operations that operates on linearized (1-D) vectors.  `vector.shuffle` operations operating on n-D (n > 1) are also converted to equivalent shuffle operations working on linearized vectors.

### CODE REVIEW

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp [hunk 1]
```diff
@@ -103,6 +105,234 @@ struct LinearizeVectorizable final
...
+    auto dstType = getTypeConverter()->convertType(shuffleOp.getType());
+    auto loc = shuffleOp.getLoc();
+    if (!dstType)
+      return rewriter.notifyMatchFailure(loc, "cannot convert type.");
+
+    if (shuffleOp.getV1VectorType().isScalable() ||
+        shuffleOp.getV2VectorType().isScalable() ||
+        dstType.cast<VectorType>().isScalable())
+      return rewriter.notifyMatchFailure(loc,
+                                         "scalable vectors are not supported.");
```
Comment (banach-space) [inline mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:242]: Could you add a test to capture this? You should be able to re-use linearlize.mlir - just add `-verify-diagnostics` to the RUN line. Same for the other pattern.

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp [hunk 2]
```diff
@@ -103,6 +105,234 @@ struct LinearizeVectorizable final
     return success();
   }
 
+private:
+  unsigned targetVectorBitWidth;
+};
+
+struct LinearizeVectorExtractStridedSlice final
```
Comment (banach-space) [inline mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:112]: Please add comments summarising this transformation. Would be good to have that for other patterns too.

File: mlir/test/Dialect/Vector/linearize.mlir [hunk 3]
```diff
@@ -153,3 +153,95 @@ func.func @test_0d_vector() -> vector<f32> {
...
+  return %0 : vector<2x2xf32>
+}
+
+// -----
+// ALL-LABEL: test_extract_strided_slice_2
+// ALL-SAME: (%[[ORIG_ARG:.*]]: vector<2x8x2xf32>) -> vector<1x4x2xf32> {
+func.func @test_extract_strided_slice_2(%arg0 : vector<2x8x2xf32>) -> vector<1x4x2xf32> {
+  // DEFAULT: %[[ARG:.*]] = vector.shape_cast %[[ORIG_ARG]] : vector<2x8x2xf32> to vector<32xf32>
+  // DEFAULT: %[[SHUFFLE:.*]] = vector.shuffle %[[ARG]], %[[ARG]]
+  // DEFAULT: [20, 21, 22, 23, 24, 25, 26, 27] : vector<32xf32>, vector<32xf32>
```
Comment (banach-space) [inline mlir/test/Dialect/Vector/linearize.mlir:186]: What are these magic numbers?

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp [hunk 4]
```diff
@@ -103,6 +105,234 @@ struct LinearizeVectorizable final
...
+    // outermost dim.
+    if (rank > 1) {
+      auto shape = shuffleOp.getV1().getType().getShape();
+      for (unsigned i = 1; i < shape.size(); i++) {
+        shuffleSliceLen *= shape[i];
+      }
+    }
+
+    auto mask = shuffleOp.getMask();
+    auto totalSize = mask.size() * shuffleSliceLen;
```
Comment (banach-space) [inline mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:265]: `totalSize` of what?

**Review by banach-space — COMMENTED** [review]
Thanks for the contribution!

I've only been able to scan this very quickly - will take another look later, but mostly looks good. In the meantime, please make sure that you follow the guidelines. In particular:
* https://mlir.llvm.org/getting_started/Contributing/#commit-messages

File: mlir/test/Dialect/Vector/linearize.mlir (hunk 3 above)
Comment (charithaintc) [reply mlir/test/Dialect/Vector/linearize.mlir:186]: Hi @banach-space, these are the shuffle indices of the generated vector.shuffle. In this case we extract a `1x4x2` slice from the `2x8x2` vector. `[20, 21, 22, 23, 24, 25, 26, 27] ` represents the indices of the linearized source vector that needs to be shuffled to construct the extracted output vector. 

Should I include some comments with an example?

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp (hunk 4 above)
Comment (charithaintc) [reply mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp]: fixed.

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp (hunk 1 above)
Comment (charithaintc) [reply mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:242]: I tried this approach. But it does not result in any error reported to standard output. I can see the error if I use `debug-only=dialect-conversion`.

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp (hunk 2 above)
Comment (charithaintc) [reply mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:115]: Added some comments describing what the rewrite pattern does.

**Comment by charithaintc** [comment]
> Thanks for the contribution!
> 
> I've only been able to scan this very quickly - will take another look later, but mostly looks good. In the meantime, please make sure that you follow the guidelines. In particular:
> 
> * https://mlir.llvm.org/getting_started/Contributing/#commit-messages

Hi @banach-space,

Thank you very much for the review. I have addressed the comments. Please let me know if there are any additional concerns.

File: mlir/test/Dialect/Vector/linearize.mlir (hunk 3 above)
Comment (Hardcode84) [reply mlir/test/Dialect/Vector/linearize.mlir:186]: I think check can be changed to `// DEFAULT-SAME:` to verify it belongs the same line with shuffle.

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp [hunk 5]
```diff
@@ -139,9 +373,19 @@ void mlir::vector::populateVectorLinearizeTypeConversionsAndLegality(
...
+                                 .cast<mlir::VectorType>()
+                                 .getRank() == 1)
+                      : true);
+        }
         return std::nullopt;
       });
 
-  patterns.add<LinearizeConstant, LinearizeVectorizable>(
+  patterns.add<LinearizeConstant, LinearizeVectorizable, LinearizeVectorShffle,
+               LinearizeVectorExtract, LinearizeVectorExtractStridedSlice>(
```
Comment (dcaballe) [inline mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:389]: Could we move this to a different `populate` method? Flattening an elementwise op is potentially a no-op but generating shuffles for extract ops could have a noticeable impact in performance

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp [hunk 6]
```diff
@@ -106,6 +108,238 @@ struct LinearizeVectorizable final
...
+  }
+
+private:
+  unsigned targetVectorBitWidth;
+};
+
+
+/// This pattern converts the vector.shuffle operation that works on nD (n > 1)
+/// vectors to a vector.shuffle operation that works on linearized vectors.
+struct LinearizeVectorShffle final
```
Comment (dcaballe) [inline mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:221]: Shffle -> Shuffle?

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp [hunk 7]
```diff
@@ -106,6 +108,238 @@ struct LinearizeVectorizable final
...
+/// vector.shuffle operation that works on a linearized vector.
+struct LinearizeVectorExtractStridedSlice final
+    : public mlir::OpConversionPattern<mlir::vector::ExtractStridedSliceOp> {
+  using OpConversionPattern::OpConversionPattern;
+  LinearizeVectorExtractStridedSlice(
+      const TypeConverter &typeConverter, MLIRContext *context,
+      unsigned targetVectBitWidth = std::numeric_limits<unsigned>::max(),
+      PatternBenefit benefit = 1)
+      : OpConversionPattern(typeConverter, context, benefit),
+        targetVectorBitWidth(targetVectBitWidth) {}
```
Comment (dcaballe) [inline mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:123]: Thanks for adding support for the target vector bitwidth!

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp [hunk 8]
```diff
@@ -106,6 +108,238 @@ struct LinearizeVectorizable final
...
+    // if kD offsets are specified for nd source vector (n > k), the granularity
+    // of the extraction is greater than 1. In this case last (n-k) dimensions
+    // form the extraction granularity. 
+    // example : 
+    //  %0 = vector.extract_strided_slice %src { offsets = [0, 0], sizes = [2, 2],
+    //     strides = [1, 1]} : vector<4x8x8xf32> to vector<2x2x8xf32>
+    // here, extraction granularity is 8.
+    int64_t extractSliceLen = 1;
+    auto n = extractOp.getSourceVectorType().getRank();
+    auto k = (int64_t)offsets.size();
```
Comment (dcaballe) [inline mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:156]: nit: we should spell out `auto` when the type is not literally redundant on the RHS.

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp [hunk 9]
```diff
@@ -106,6 +108,238 @@ struct LinearizeVectorizable final
...
+    if (n > k) {
+      for (unsigned i = 0; i < n - k; i++) {
+        extractSliceLen *= extractOp.getSourceVectorType().getShape()[i + k];
+      }
+    }
+    // get total number of extracted slices
```
Comment (dcaballe) [inline mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:162]: nit: we should use capital letters and periods in comments per coding guidelines.

**Review by dcaballe — COMMENTED** [review]
I did a first pass. LG! A few comments

**Comment by charithaintc** [comment]
> I did a first pass. LG! A few comments

Hi @dcaballe,

Thanks for the review. I have addressed all the comments.

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp (hunk 5 above)
Comment (charithaintc) [reply mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp]: I added a new populate method called `populateVectorLinearizeToShuffleRewritePatterns`. If you have a better name suggestion for this please let me know.

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp (hunk 6 above)
Comment (charithaintc) [reply mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp]: fixed.

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp (hunk 8 above)
Comment (charithaintc) [reply mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp]: fixed.

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp (hunk 9 above)
Comment (charithaintc) [reply mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp]: fixed all comments.

File: mlir/test/Dialect/Vector/linearize.mlir (hunk 3 above)
Comment (charithaintc) [reply mlir/test/Dialect/Vector/linearize.mlir]: added `// *-SAME` to all lines with shuffle indices.

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp [hunk 10]
```diff
@@ -145,3 +378,24 @@ void mlir::vector::populateVectorLinearizeTypeConversionsAndLegality(
   patterns.add<LinearizeConstant, LinearizeVectorizable>(
       typeConverter, patterns.getContext(), targetBitWidth);
 }
+
+void mlir::vector::populateVectorLinearizeToShuffleRewritePatterns(
+    TypeConverter &typeConverter, RewritePatternSet &patterns,
+    ConversionTarget &target, unsigned int targetBitWidth) {
+  target.markUnknownOpDynamicallyLegal(
```
Comment (Hardcode84) [inline mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:385]: I think you can just do `target.addDynamicallyLegalOp<vector::ShuffleOp>(...)`

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp [hunk 11]
```diff
@@ -103,6 +106,236 @@ struct LinearizeVectorizable final
...
+      const TypeConverter &typeConverter, MLIRContext *context,
+      unsigned targetVectBitWidth = std::numeric_limits<unsigned>::max(),
+      PatternBenefit benefit = 1)
+      : OpConversionPattern(typeConverter, context, benefit),
+        targetVectorBitWidth(targetVectBitWidth) {}
+
+  LogicalResult
+  matchAndRewrite(vector::ExtractStridedSliceOp extractOp, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    auto dstType = getTypeConverter()->convertType(extractOp.getType());
```
Comment (banach-space) [inline mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:128]: Why is this needed?

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp [hunk 12]
```diff
@@ -103,6 +106,236 @@ struct LinearizeVectorizable final
...
+    auto loc = extractOp.getLoc();
+    if (!dstType)
+      return rewriter.notifyMatchFailure(loc, "cannot convert type.");
+    if (extractOp.getVector().getType().isScalable() ||
+        dstType.cast<VectorType>().isScalable())
+      return rewriter.notifyMatchFailure(loc,
+                                         "scalable vectors are not supported.");
+    if (!isLessThanTargetBitWidth(extractOp, targetVectorBitWidth))
+      return rewriter.notifyMatchFailure(
+          extractOp, "Can't flatten since targetBitWidth <= OpSize");
```
Comment (banach-space) [inline mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:138]: This would correspond to `BW-0` in your tests, right?

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp [hunk 13]
```diff
@@ -103,6 +106,236 @@ struct LinearizeVectorizable final
...
+
+    auto offsets = extractOp.getOffsets().getValue();
+    auto sizes = extractOp.getSizes().getValue();
+    auto strides = extractOp.getStrides().getValue();
```
Comment (banach-space) [inline mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:142]:
```suggestion
    auto offsets = extractOp.getOffsets();
    auto sizes = extractOp.getSizes();
    auto strides = extractOp.getStrides();
```

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp [hunk 14]
```diff
@@ -103,6 +106,236 @@ struct LinearizeVectorizable final
...
+          extractOp, "Strided slice with stride != 1 is not supported.");
+    Value srcVector = adaptor.getVector();
+    // If kD offsets are specified for nd source vector (n > k), the granularity
+    // of the extraction is greater than 1. In this case last (n-k) dimensions
+    // form the extraction granularity.
+    // example :
+    //  %0 = vector.extract_strided_slice %src { offsets = [0, 0], sizes = [2,
+    //  2],
+    //     strides = [1, 1]} : vector<4x8x8xf32> to vector<2x2x8xf32>
+    // here, extraction granularity is 8.
```
Comment (banach-space) [inline mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:154]:
```suggestion
    // If kD offsets are specified for nD source vector (n > k), the granularity
    // of the extraction is greater than 1. In this case last (n-k) dimensions
    // form the extraction granularity.
    // example :
    //  vector.extract_strided_slice %s {
    //        offsets = [0, 0], sizes = [2, 2], strides = [1, 1]} : 
    //        vector<4x8x8xf32> to vector<2x2x8xf32>
    // here, extraction granularity is 8.
```

1. Please use consistent style in your comments (capitalisation).
2. Aavoid "random" line wrapping (e.g. in `sizes = [2, 2]`) -  that makes it hard to read.
3. This comment defines "granularity", but the code that follows computes `extractSliceLen`. Please make it more consistent with the implementation.

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp [hunk 15]
```diff
@@ -103,6 +106,236 @@ struct LinearizeVectorizable final
...
+    int64_t extractSliceLen = 1;
```
Comment (banach-space) [inline mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:155]: This represents "flattened slice len", right? Why not `flattenedSliceLen`? Naming is hard 🤷🏻

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp [hunk 16]
```diff
@@ -103,6 +106,236 @@ struct LinearizeVectorizable final
...
+    auto n = extractOp.getSourceVectorType().getRank();
+    int64_t k = (int64_t)offsets.size();
+    if (n > k) {
```
Comment (banach-space) [inline mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:158]: Is this `if` is needed here? `k` is guaranteed to be ` <= n`, right?

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp [hunk 17]
```diff
@@ -103,6 +106,236 @@ struct LinearizeVectorizable final
...
+      for (unsigned i = 0; i < n - k; i++) {
+        extractSliceLen *= extractOp.getSourceVectorType().getShape()[i + k];
+      }
+    }
+    // Get total number of extracted slices.
+    int64_t nExtractedSlices = 1;
+    for (auto size : sizes) {
+      nExtractedSlices *= size.cast<IntegerAttr>().getInt();
+    }
```
Comment (banach-space) [inline mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:167]: Could you try using `computeSuffixProduct` instead?

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp (hunk 1 above)
Comment (charithaintc) [reply mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp]: Good point! #88733 is merged now. so replaced with assert.

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp (hunk 2 above)
Comment (charithaintc) [reply mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:124]: Thanks for pointing this out. I tried my best to add some example based comments for each pattern.

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp (hunk 10 above)
Comment (charithaintc) [reply mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp]: fixed.

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp (hunk 11 above)
Comment (charithaintc) [reply mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:138]: yes. this is not needed. I removed the `dstType` checks.

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp (hunk 12 above)
Comment (charithaintc) [reply mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:147]: yes.

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp (hunk 13 above)
Comment (charithaintc) [reply mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:151]: fixed.

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp (hunk 14 above)
Comment (charithaintc) [reply mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp]: renamed to `extractGranularitySize`. I think the work granularity captures the meaning well. so I like to keep it.

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp (hunk 15 above)
Comment (charithaintc) [reply mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:164]: addressed above. basically the extraction can happen only at first k dimensions of an n-D vectors. the remaining dimensions are extracted in full. so the last n-k dimensions form the extraction granularity. this is the meaning I wanted to capture in the naming.

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp (hunk 16 above)
Comment (charithaintc) [reply mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:167]: addressed above. n > k is true for the following example. 
``` 
%0 = vector.extract_strided_slice %arg0 { offsets = [1, 2], strides = [1, 1], sizes = [1, 4] }
    : vector<2x8x2xf32> to vector<1x4x2xf32>
```
In this case, only first 2 dims are extracted using the offsets, sizes etc. last dim is extracted in full. 
https://mlir.llvm.org/docs/Dialects/Vector/#vectorextract_strided_slice-vectorextractstridedsliceop

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp (hunk 17 above)
Comment (charithaintc) [reply mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:176]: tried this. but this require constructing an `ArrayRef<int64_t>`. So I decided to use `llvm::for_each`.

**Comment by charithaintc** [comment]
Hi, @banach-space, @dcaballe, @Hardcode84. I think I have addressed all the comments. Can you please review the modifications and/or approve.

Thanks!

**Comment by Hardcode84** [comment]
LGTM, but wait for other reviewers

File: mlir/include/mlir/Dialect/Vector/Transforms/VectorRewritePatterns.h [hunk 18]
```diff
@@ -389,6 +389,12 @@ void populateVectorLinearizeTypeConversionsAndLegality(
     TypeConverter &typeConverter, RewritePatternSet &patterns,
     ConversionTarget &target, unsigned targetBitWidth);
 
+/// Populates patterns for linearizing ND (N >= 2) vector operations to 1D
+/// vector shuffle operations.
+void populateVectorLinearizeToShuffleRewritePatterns(
```
Comment (dcaballe) [inline mlir/include/mlir/Dialect/Vector/Transforms/VectorRewritePatterns.h:394]: Extract ops are "shuffle-like" ops so perhaps we can call this `populateVectorLinearizeShuffleLikeOpsPatterns` or similar

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp (hunk 11 above)
Comment (dcaballe) [reply mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:139]: nit: spell out `auto`

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp [hunk 19]
```diff
@@ -103,6 +108,252 @@ struct LinearizeVectorizable final
...
+    // If kD offsets are specified for nd source vector (n > k), the granularity
+    // of the extraction is greater than 1. In this case last (n-k) dimensions
+    // form the extraction granularity.
+    // Example :
+    //  vector.extract_strided_slice %src {
+    //      offsets = [0, 0], sizes = [2, 2], strides = [1, 1]} :
+    //      vector<4x8x8xf32> to vector<2x2x8xf32>
+    // Here, extraction granularity is 8.
+    int64_t extractGranularitySize = 1;
+    auto n = extractOp.getSourceVectorType().getRank();
```
Comment (dcaballe) [inline mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:163]: nit: please, spell out the type when it's not literally redundant in the statement.

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp [hunk 20]
```diff
@@ -103,6 +108,252 @@ struct LinearizeVectorizable final
...
+    int64_t k = (int64_t)offsets.size();
+    if (n > k) {
+      for (unsigned i = 0; i < n - k; i++) {
```
Comment (dcaballe) [inline mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:166]: nit: use pre-increment per coding guidelines

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp [hunk 21]
```diff
@@ -103,6 +108,252 @@ struct LinearizeVectorizable final
...
+        extractGranularitySize *=
+            extractOp.getSourceVectorType().getShape()[i + k];
+      }
+    }
+    // Get total number of extracted slices.
+    int64_t nExtractedSlices = 1;
+    llvm::for_each(sizes, [&](Attribute size) {
```
Comment (dcaballe) [inline mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:173]: `for (Attribute size : sizes) {...}` ?

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp [hunk 22]
```diff
@@ -103,6 +108,252 @@ struct LinearizeVectorizable final
...
+    assert(!(shuffleOp.getV1VectorType().isScalable() ||
+             shuffleOp.getV2VectorType().isScalable() ||
+             dstType.cast<VectorType>().isScalable()) &&
+           "scalable vectors are not supported.");
+    if (!isLessThanTargetBitWidth(shuffleOp, targetVectorBitWidth))
+      return rewriter.notifyMatchFailure(
+          shuffleOp, "Can't flatten since targetBitWidth <= OpSize");
+
+    auto vec1 = adaptor.getV1();
+    auto vec2 = adaptor.getV2();
```
Comment (dcaballe) [inline mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:263]: nit: spell out `auto`

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp [hunk 23]
```diff
@@ -103,6 +108,252 @@ struct LinearizeVectorizable final
...
+    int shuffleSliceLen = 1;
+    int rank = shuffleOp.getV1().getType().getRank();
+
+    // If rank > 1, we need to do the shuffle in the granularity of slices
+    // instead of scalars. Size of the slice is equal to the rank-1 innermost
+    // dims. Mask of the shuffle op specifies which slice to take from the
+    // outermost dim.
+    if (rank > 1) {
+      auto shape = shuffleOp.getV1().getType().getShape();
+      for (unsigned i = 1; i < shape.size(); i++) {
```
Comment (dcaballe) [inline mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:273]: pre-increment

**Review by dcaballe — APPROVED** [review]
LGTM (module some nits that you can address before landing)

File: mlir/include/mlir/Dialect/Vector/Transforms/VectorRewritePatterns.h (hunk 18 above)
Comment (charithaintc) [reply mlir/include/mlir/Dialect/Vector/Transforms/VectorRewritePatterns.h:394]: renamed to `populateVectorLinearizeShuffleLikeOpsPatterns`

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp (hunk 19 above)
Comment (charithaintc) [reply mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:163]: fixed!

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp (hunk 20 above)
Comment (charithaintc) [reply mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:166]: fixed.

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp (hunk 22 above)
Comment (charithaintc) [reply mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:263]: fixed.

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp (hunk 21 above)
Comment (charithaintc) [reply mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:173]: @banach-space recommended using `computeSuffixProduct`. But that does not take AttrArray as argument. So I am not sure to use it for this case. Changed back to a for loop for now.

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp (hunk 23 above)
Comment (charithaintc) [reply mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:273]: fixed.

File: mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp (hunk 16 above)
Comment (banach-space) [reply mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:167]: Right, so there are 2 cases:
* `k < n`
* `k = n`

right? I think that you can get rid of `if` by doing something like:
```cpp
int64_t outRank = (int64_t)offsets.size();
int64_t k = outRank;
while (k < n) {
  extractGranularitySize *=
            extractOp.getSourceVectorType().getShape()[outRank + (n-k)];
  ++k;
}
```

This way there's only one case. Also, I'd avoid variable names like `k` and `n` for anything other than loop iterators. Perhaps `numKDims` and `numMDims`? Or `inputRank`/`outputRank`? Or `kD`/`nD`? The last suggestion would be consistent with comments and already a nice improvement :)

**Review by banach-space — APPROVED** [review]
LGTM, left a few final suggestions, thanks!

**Comment by charithaintc** [comment]
Hi @Hardcode84, @banach-space, @dcaballe,

Thanks for the insightful reviews. This is my first PR in llvm-project and I learned a lot 😃 👍 

Can you please help me with the merge process.

Thanks!
